src/Microsoft.Android.Sdk.ILLink/LinkContextExtensions.cs

```diff
@@ -0,0 +1,161 @@
+#nullable enable
+
+using Mono.Cecil;
+using Mono.Collections.Generic;
+using Mono.Linker;
+using System.Collections.Generic;
+
+namespace Microsoft.Android.Sdk.ILLink
+{
+	public static class LinkContextExtensions
+	{
+		public static MethodDefinition GetBaseDefinition (this LinkContext context, MethodDefinition method)
+		{
+			if (method.IsStatic || method.IsNewSlot || !method.IsVirtual)
+				return method;
+
+			foreach (var baseType in context.GetBaseTypes (method.DeclaringType)) {
+				foreach (var m in baseType.Methods) {
+					if (!m.IsConstructor &&
+							m.Name == method.Name &&
+							(m.IsVirtual || m.IsAbstract) &&
+							context.AreParametersCompatibleWith (m.Parameters, method.Parameters)) {
```

Comment:
I do not believe that this check is sufficient; see also commit f96fcf93e157472072576bcc0a8698302899e8cf.

In particular, for interface members, the `.override` IL directive is used to help mark method overrides, and IL permits *contravariance*:

```cil
// Metadata version: v4.0.30319
.assembly extern mscorlib
{
  .ver 4:0:0:0
}
.assembly override
{
}


.class interface private abstract auto ansi IExample
{
  .method public hidebysig newslot abstract virtual 
          instance void  IM(string s) cil managed
  {
  } // end of method IExample::IM

} // end of class IExample

.class private auto ansi beforefieldinit Base
       extends [mscorlib]System.Object
       implements IExample
{
  .method private hidebysig newslot virtual final 
          instance void  IExample.IM(string s) cil managed
  {
    .override IExample::IM
    ldstr      "Base::IM"
    call       void [mscorlib]System.Console::WriteLine(string)
    ret
  } // end of method Base::IExample.IM

  .method public hidebysig specialname rtspecialname 
          instance void  .ctor() cil managed
  {
    ldarg.0
    call       instance void [mscorlib]System.Object::.ctor()
    ret
  } // end of method Base::.ctor

} // end of class Base

.class private auto ansi beforefieldinit Derived
       extends Base
       implements IExample
{
  .method private hidebysig newslot virtual final 
          instance void  IExample.IM(object s) cil managed
  {
    .override method instance void IExample::IM(string s)
    ldstr      "Derived::IM"
    call       void [mscorlib]System.Console::WriteLine(string)
    ret
  } // end of method Derived::IExample.IM

  .method public hidebysig specialname rtspecialname 
          instance void  .ctor() cil managed
  {
    ldarg.0
    call       instance void Base::.ctor()
    ret
  } // end of method Derived::.ctor

} // end of class Derived

.class private auto ansi beforefieldinit App
       extends [mscorlib]System.Object
{
  .method public hidebysig static void  Main() cil managed
  {
    .entrypoint
    .maxstack  2
    .locals init (class Base V_0)
    nop
    newobj     instance void Derived::.ctor()
    stloc.0
    ldloc.0
    ldnull
    callvirt   instance void IExample::IM(string)
    nop
    ret
  } // end of method App::Main

} // end of class App
```

This is equivalent to:

```csharp
interface IExample {
    void IM(string s);
}
class Base : IExample {
    void IExample::IM(string s) { Console.WriteLine("Base::IM"); }
}
class Derived : Base, IExample {
    // not valid C#; parameter contravariance
    void IExample::IM(object s) { Console.WriteLine("Derived::IM"); }
}
class App {
    public static void Main() {
        Base b = new Derived();
        ((IExample)b).IM(null);
    }
}
```

This is valid IL, but `context.AreParametersCompatibleWith()` won't view this as a method override.

*Instead*, you should use the `MethodDefinition.Overrides` collection: https://github.com/xamarin/xamarin-android/blob/92efdccb686454be32d742b8ca9f6a166a7c1808/src/Xamarin.Android.Build.Tasks/Linker/MonoDroid.Tuner/FixAbstractMethodsStep.cs#L160-L170

Unfortunately, `MethodDefinition.Overrides` only helps with interface methods.

Reply:
I see what you're saying - I am wondering whether it's more important to be consistent with java.interop's `GetBaseDefinition` or to be correct in this case. I will leave it as-is until we decide what to do with the duplicated logic.

Reply:
@jonpryor now that I've removed the copy of these extension methods - is it worth fixing this in java.interop?

Reply:
@sbomer asked:

> is it worth fixing this in java.interop?

Probably yes?  The problem is that I considered doing this fix as part of xamarin/java.interop@2573dc8c84fd4eb68e75bcae73912c26f4942356, but wasn't quite sure how to actually *do* it, based on the "scope" and information available.

In particular, `FixAbstractMethodStep.IsInOverrides()` (linked above) *knows* that it has an interface method `iMethod` and a type method `tMethod`.  It thus makes sense to ask "is `iMethod` one of the methods that `tMethod` overrides?

`FixAbstractMethodStep` can do this because it likewise knows when it's dealing with an interface, and thus *has* `iMethod`.

`MethodDefinitionRocks` in Java.Interop has *none* of that additional context, at least not structured in the same way as `FixAbstractMethodStep` thinks about things.

Consequently, while we (probably?) should fix this in Java.Interop, it wasn't a "trivial minimal thought" fix, so I skipped it.

I should file a bug about that, though…?

That said, the primary/original use case for `MethodDefinitionRocks` is in [`JavaCallableWrapperGenerator.cs`](https://github.com/xamarin/java.interop/blob/main/src/Java.Interop.Tools.JavaCallableWrappers/Java.Interop.Tools.JavaCallableWrappers/JavaCallableWrapperGenerator.cs), which is responsible for generating a Java "proxy" for a managed type, and it works by grabbing the closure of all interface methods *plus* all "override" methods to find the set of methods with a `[Register]` attribute, and if there's overlap between these approaches -- because a base class implements an interface via virtual method, so this lookup would find the interface method and virtual method separately -- then that's fine, we'll appropriately skip the "duplicate".

Thus, *in the context of Java.Interop*, this isn't a problem.

*In the context of the linker*… I have no idea.

Reply:
I don't know how important this is for the android linker steps either. Not sure if this is helpful, but I filed https://github.com/xamarin/xamarin-android/issues/5974 to track these concerns. I know the built-in linker steps have had issues related to this as well, so it is probably worth trying to get this right in the linker, with plenty of tests, etc. and then being consistent about it in xamarin-android.